cordova-lib/src/cordova/platform_metadata.js

```diff
@@ -0,0 +1,102 @@
+    software distributed under the License is distributed on an
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+*/
+
+
+var path         = require('path'),
+    cordova_util = require('./util'),
+    fs           = require('fs'),
+    Q            = require('q'),
+    shelljs      = require('shelljs');
+
+
+// Retrieves the platforms and their versions from the platforms.json file
+// Returns an array of {platform: platform, version: version} ...
+// ... where version could be '3.4.0', '/path/to/platform' or 'git://...'
+function getVersions(projectRoot) {
+    var platformsDir = path.join(projectRoot, 'platforms');
+    var platformsJsonFile = path.join(platformsDir, 'platforms.json');
+
+    // If the platforms.json file doesn't exist, retrieve versions from platforms installed on the filesystem...
+    // ...Note that in this case, we won't be able to know what source(folder, git-url) the platform came from, we'll just use versions
+    return getPlatVersionsFromFile(platformsJsonFile).fail(function(){
+        return getPlatVersionsFromFileSystem(projectRoot);
+    });
+}
+
+// Returns a promise
+function getPlatVersionsFromFile(platformsJsonFile){
+    var platformData = require(platformsJsonFile);
+    var platformVersions = [];
+    
+    platformVersions = Object.keys(platformData).map(function(p){
+        return {platform: p, version: platformData[p]};
+    });
+
+    return Q(platformVersions);
+}
+
+// Returns a promise
+function getPlatVersionsFromFileSystem(projectRoot){
+    var platformVersions = [];
+    var platforms_on_fs = cordova_util.listPlatforms(projectRoot);
+    platforms_on_fs.forEach(function(platform){
+        var script = path.join(projectRoot, 'platforms', platform, 'cordova', 'version');
+        
+        // Use version script to retrieve installed version of the platform
+        var version = shelljs.exec(script).output;
+
+        // clean the version we get back from the script
+        // This is necessary because the version script uses console.log to pass back
+        // the version. Using console.log ends up adding additional line breaks/newlines to the value returned. 
+        // ToDO: version scripts should be refactored to not use console.log()
+        var versionCleaned = version.replace(/\r?\n|\r/g, '');
+        platformVersions.push({platform: platform, version: versionCleaned});    
+    });
+    
+    return Q(platformVersions); 
+}
+
+// Saves platform@version into platforms.json
+function save(projectRoot, platform, version) {
+    var platformsDir = path.join(projectRoot, 'platforms');
+    var platformJsonFile = path.join(platformsDir, 'platforms.json');
+    if(!fs.existsSync(platformJsonFile)){
+        fs.writeFileSync(platformJsonFile, JSON.stringify({}, null, 4), 'utf-8');
+    }
+
+    var data = require(platformJsonFile);
```

Comment:
It's generally not very flexible to read and write a file in the same method. Better to have clients explicitly read, explicitly mutate (in memory), then explicitly write.